### PR TITLE
Javalab: Run correct code in preview mode

### PR DESCRIPTION
**Pull request by sanchitmalhotra126**
Ensures that we run the correct in preview mode (when viewing another version of a project). With this change, we'll use the sources currently on the page, as opposed to the sources saved on the backend in S3 (like exemplar mode).

## Links

JIRA: https://codedotorg.atlassian.net/browse/JAVA-666

## Testing story

Tested locally with multiple versions, ensuring the correct code was executed.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/javalab/Javalab.js [hunk 1]
```diff
@@ -388,7 +388,13 @@ Javalab.prototype.executeJavabuilder = function(executionType) {
   );
 
   let connectToJavabuilder;
-  if (this.isEditingExemplar || this.isViewingExemplar) {
+  // If in exemplar or preview mode, use the sources currently on the page
+  // (as opposed to the sources saved on the backend).
+  if (
+    this.isEditingExemplar ||
+    this.isViewingExemplar ||
+    getStore().getState().pageConstants.isReadOnlyWorkspace
```
Comment (molly-moen) [inline apps/src/javalab/Javalab.js:396]: out of scope for this pr, but I am wondering if we should ever use the sources saved on the backend? We probably always want to run what is showing to the user, which is often what is also saved on the backend. But if they aren't the same we should go with what the user sees.

**Review by molly-moen — APPROVED** [review]

File: apps/src/javalab/Javalab.js (hunk 1 above)
Comment (sanchitmalhotra126) [reply apps/src/javalab/Javalab.js:396]: Yeah good question - IIRC we considered it while designing the decoupled javabuilder solution, but ultimately decided not to, due to us wanting to hide validation code from the user. But looks like we supply validation from the client now as well, so not sure if that's still a concern...
Edit: we only send validation in start mode, so maybe it's still a valid concern?
Also edit: regardless though, that shouldn't affect the contents of the sources we send

Comment (molly-moen) [reply apps/src/javalab/Javalab.js:396]: oh right that's why. I think sources could always come from the client but it's not a high priority

Comment (sanchitmalhotra126) [reply apps/src/javalab/Javalab.js:396]: Yeah agreed - I'll add a task

Comment (jmkulwik) [reply apps/src/javalab/Javalab.js:396]: If I'm remembering right, we decided to use sources from the backend for a few reasons:
* validation code, as mentioned
* too large of a payload - student projects in App Lab regularly grow to low thousands of lines of code, we suspected the same would happen in Java Lab. With the non-decoupled version of Javabuilder, we would have quickly surpassed payload limits.
* consistency between the backend and the code that is run. This is an issue we run into quite often in certain labs, such as Web Lab. Sometimes an issue happens with the PUT so the code that is executed is different from what's saved on S3. Then, when the student reloads the page, they have a different experience from the last time they were viewing their code.
^ I think so long as we wait for the save to complete successfully before running the code, that last item shouldn't be an issue.

Meta point: this is moreso me thinking out loud and providing context. I was initially a bit concerned about this change, but couldn't put my finger on why. Writing these out made me realize it was bullet 2 (already solved for) and bullet 3 (unlikely to be an issue for us). So this ends up being a less useful comment, but I'll go ahead and leave it to provide context in the future if needed.

**Pull request merged**